packages/counters.lua

```diff
@@ -52,91 +52,103 @@ SILE.formatCounter = function(options)
 
-local _init = function (c)
-  if not(SILE.scratch.counters[c]) then
-    SILE.scratch.counters[c] = { value= 0, display= "arabic" }
+local function getCounter(id)
+  local counter = SILE.scratch.counters[id]
+  if not counter then
+    counter = {}
+    SILE.scratch.counters[id] = { value= 0, display= "arabic", format = SILE.formatCounter }
   end
+  return counter
 end
 
-SILE.registerCommand("increment-counter", function (options,content)
-  local c = options.id; _init(c)
+SILE.registerCommand("increment-counter", function (options, _)
+  local counter = getCounter(options.id)
   if (options["set-to"]) then
-    SILE.scratch.counters[c].value = tonumber(options["set-to"])
+    counter.value = tonumber(options["set-to"])
   else
-    SILE.scratch.counters[c].value = SILE.scratch.counters[c].value + 1
+    counter.value = counter.value + 1
   end
-  if options.display then SILE.scratch.counters[c].display = options.display end
+  if options.display then counter.display = options.display end
 end, "Increments the counter named by the <id> option")
 
-SILE.registerCommand("set-counter", function (options, content)
-  local c = options.id; _init(c)
-  if options.value then SILE.scratch.counters[c].value = tonumber(options.value) end
-  if options.display then SILE.scratch.counters[c].display = options.display end
+SILE.registerCommand("set-counter", function (options, _)
+  local counter = getCounter(options.id)
+  if options.value then counter.value = tonumber(options.value) end
+  if options.display then counter.display = options.display end
 end, "Sets the counter named by the <id> option to <value>; sets its display type (roman/Roman/arabic) to type <display>.")
 
 
-SILE.registerCommand("show-counter", function (options, content)
-  local c = options.id; _init(c)
-  if options.display then SILE.scratch.counters[c].display = options.display end
-  SILE.typesetter:setpar(SILE.formatCounter(SILE.scratch.counters[c]))
+SILE.registerCommand("show-counter", function (options, _)
+  local counter = getCounter(options.id)
+  if options.display then counter.display = options.display end
+  SILE.typesetter:setpar(SILE.formatCounter(counter))
 end, "Outputs the value of counter <id>, optionally displaying it with the <display> format.")
 
-local _initml = function (c)
-  if not(SILE.scratch.counters[c]) then
-    SILE.scratch.counters[c] = { value= {0}, display= {"arabic"} }
+SILE.formatMultilevelCounter = function(counter, options)
+  local maxlevel = options and options.level or #counter.value
+  local minlevel = options and options.minlevel or 1
+  local out = {}
+  for x = minlevel, maxlevel do
+    out[x - minlevel + 1] = SILE.formatCounter({ display = counter.display[x], value = counter.value[x] })
   end
+  return table.concat( out, "." )
 end
 
-SILE.registerCommand("increment-multilevel-counter", function (options, content)
-  local c = options.id; _initml(c)
-  local this = SILE.scratch.counters[c]
+local function getMultilevelCounter(id)
```

Comment:
This isn't the fault of this commit, but it's always bugged me how much duplication there was between counters and multilevel counters. I think we need to redo the backend for all counters to be the same (multilevel) and default everything to acting on the top (or only) level, and regular counters could just format that level only.

Reply:
Sounds good. I propose to do that in a different PR, to keep it lean. I can do it after this gets merged.

Reply:
Okay.